relay/relay.go

```diff
@@ -43,28 +46,189 @@ func UnmarshalSpec(id graphql.ID, v interface{}) error {
+}
+
+// NewRequestOptions Parses an http.Request into GraphQL request options struct
+func NewRequestOptions(r *http.Request) *RequestOptions {
+	if reqOpt := getFromForm(r.URL.Query()); reqOpt != nil {
+		return reqOpt
 	}
 
-	response := h.Schema.Exec(r.Context(), params.Query, params.OperationName, params.Variables)
-	responseJSON, err := json.Marshal(response)
-	if err != nil {
-		http.Error(w, err.Error(), http.StatusInternalServerError)
-		return
+	if r.Method != "POST" {
+		return &RequestOptions{}
+	}
+
+	if r.Body == nil {
+		return &RequestOptions{}
+	}
+
+	// TODO: improve Content-Type handling
+	contentTypeStr := r.Header.Get("Content-Type")
+	contentTypeTokens := strings.Split(contentTypeStr, ";")
+	contentType := contentTypeTokens[0]
+
+	switch contentType {
+	case ContentTypeGraphQL:
+		body, err := ioutil.ReadAll(r.Body)
+		if err != nil {
+			return &RequestOptions{}
+		}
+		return &RequestOptions{
+			Query: string(body),
+		}
+	case ContentTypeFormURLEncoded:
+		if err := r.ParseForm(); err != nil {
+			return &RequestOptions{}
+		}
+
+		if reqOpt := getFromForm(r.PostForm); reqOpt != nil {
+			return reqOpt
+		}
+
+		return &RequestOptions{}
+
+	case ContentTypeJSON:
+		fallthrough
+	default:
+		var opts RequestOptions
+		body, err := ioutil.ReadAll(r.Body)
+		if err != nil {
+			return &opts
+		}
+		err = json.Unmarshal(body, &opts)
+		if err != nil {
+			// Probably `variables` was sent as a string instead of an object.
+			// So, we try to be polite and try to parse that as a JSON string
+			var optsCompatible requestOptionsCompatibility
+			json.Unmarshal(body, &optsCompatible)
+			json.Unmarshal([]byte(optsCompatible.Variables), &opts.Variables)
+		}
+		return &opts
+	}
+}
+
+type param struct {
+	Schema          *graphql.Schema
+	RequestString   string
+	VariablesValues map[string]interface{}
+	OperationName   string
+	Context         context.Context
+}
+
+// ContextHandler provides an entrypoint into executing graphQL queries with a
+// user-provided context.
+func (h *Handler) ContextHandler(ctx context.Context, w http.ResponseWriter, r *http.Request) {
+	// Gets Query
+	opts := NewRequestOptions(r)
+	params := param{
+		Schema:          h.Schema,
+		RequestString:   opts.Query,
+		VariablesValues: opts.Variables,
+		OperationName:   opts.OperationName,
+		Context:         ctx,
 	}
+	response := h.Schema.Exec(ctx, params.RequestString, params.OperationName, params.VariablesValues)
+
+	if h.graphiql {
+		acceptHeader := r.Header.Get("Accept")
+		_, raw := r.URL.Query()["raw"]
+		if !raw && !strings.Contains(acceptHeader, ContentTypeJSON) && strings.Contains(acceptHeader, "text/html") {
+			renderGraphiQL(w, params)
+			return
+		}
+	}
+
+	// Use proper JSON Header
+	w.Header().Add("Content-Type", "application/json; charset=utf-8")
 
-	w.Header().Set("Content-Type", "application/json")
-	w.Write(responseJSON)
+	if h.pretty {
+		w.WriteHeader(http.StatusOK)
+		responseJSON, err := json.MarshalIndent(response, "", "\t")
```

Comment:
Why would this handler pretty print? GraphiQL + many browser plugins + `jq` already do the pretty printing for you. I guess this handler is to be consumed by other apps only, no?

Reply:
Roger that! Getting rid of it!